File: barter-rs/src/strategy/glft.py
from numba import njit
from hftbacktest import BUY, SELL

import numpy as np

@njit
def measure_trading_intensity_and_volatility(hbt):
    arrival_depth = np.full(10_000_000, np.nan, np.float64)
    mid_price_chg = np.full(10_000_000, np.nan, np.float64)

    t = 0
    prev_mid_price_tick = np.nan
    mid_price_tick = np.nan
    
    # Checks every 100 milliseconds.
    while hbt.elapse(100_000):
        #--------------------------------------------------------
        # Records market order's arrival depth from the mid-price.
        if not np.isnan(mid_price_tick):
            depth = -np.inf
            for trade in hbt.last_trades:
                side = trade[3]
                trade_price_tick = trade[4] / hbt.tick_size
                
                if side == BUY:
                    depth = np.nanmax([trade_price_tick - mid_price_tick, depth])
                else:
                    depth = np.nanmax([mid_price_tick - trade_price_tick, depth])
            arrival_depth[t] = depth
        
        hbt.clear_last_trades()
        
        prev_mid_price_tick = mid_price_tick
        mid_price_tick = (hbt.best_bid_tick + hbt.best_ask_tick) / 2.0
        
        # Records the mid-price change for volatility calculation.
        mid_price_chg[t] = mid_price_tick - prev_mid_price_tick
        
        t += 1
        if t >= len(arrival_depth) or t >= len(mid_price_chg):
            raise Exception
    return arrival_depth[:t], mid_price_chg[:t]
  
@njit
def measure_trading_intensity(order_arrival_depth, out):
    max_tick = 0
    for depth in order_arrival_depth:
        if not np.isfinite(depth):
            continue
        
        # Sets the tick index to 0 for the nearest possible best price 
        # as the order arrival depth in ticks is measured from the mid-price
        tick = round(depth / .5) - 1
        
        # In a fast-moving market, buy trades can occur below the mid-price (and vice versa for sell trades) 
        # since the mid-price is measured in a previous time-step; 
        # however, to simplify the problem, we will exclude those cases.
        if tick < 0 or tick >= len(out):
            continue
        
        # All of our possible quotes within the order arrival depth, 
        # excluding those at the same price, are considered executed.
        out[:tick] += 1
        
        max_tick = max(max_tick, tick)
    return out[:max_tick]
  
def plot():
    tmp = np.zeros(500, np.float64) 

    # Measures trading intensity (lambda) for the first 10-minute window.
    lambda_ = measure_trading_intensity(arrival_depth[:6_000], tmp)

    # Since it is measured for a 10-minute window, divide by 600 to convert it to per second.
    lambda_ /= 600

    # Creates ticks from the mid-price.
    ticks = np.arange(len(lambda_)) + .5
    
    y = np.log(lambda_)
    k_, logA = linear_regression(ticks, y)
    A = np.exp(logA)
    k = -k_

    print('A={}, k={}'.format(A, k))
    
@njit
def linear_regression(x, y):
    sx = np.sum(x)
    sy = np.sum(y)
    sx2 = np.sum(x ** 2)
    sxy = np.sum(x * y)
    w = len(x)
    slope = (w * sxy - sx * sy) / (w * sx2 - sx**2)
    intercept = (sy - slope * sx) / w
    return slope, intercept
  
@njit
def compute_coeff(xi, gamma, delta, A, k):
    inv_k = np.divide(1, k)
    c1 = 1 / (xi * delta) * np.log(1 + xi * delta * inv_k)
    c2 = np.sqrt(np.divide(gamma, 2 * A * delta * k) * ((1 + xi * delta * inv_k) ** (k / (xi * delta) + 1)))
    return c1, c2
  
@njit
def rest(): 
    #--------------------------------------------------------
    # Calibrates A, k and calculates the market volatility.
    
    # Updates A, k, and the volatility every 5-sec.
    if t % 50 == 0:
        # Window size is 10-minute.
        if t >= 6_000 - 1:
            # Calibrates A, k
            tmp[:] = 0
            lambda_ = measure_trading_intensity(arrival_depth[t + 1 - 6_000:t + 1], tmp)
            lambda_ = lambda_[:70] / 600
            x = ticks[:len(lambda_)]
            y = np.log(lambda_)
            k_, logA = linear_regression(x, y)
            A = np.exp(logA)
            k = -k_
        
            # Updates the volatility.
            volatility = np.nanstd(mid_price_chg[t + 1 - 6_000:t + 1]) * np.sqrt(10)

    #--------------------------------------------------------
    # Computes bid price and ask price.

    c1, c2 = compute_coeff(gamma, gamma, delta, A, k)
    
    half_spread = c1 + delta / 2 * c2 * volatility
    skew = c2 * volatility
    
    bid_depth = half_spread + skew * hbt.position
    ask_depth = half_spread - skew * hbt.position

    bid_price = min(np.round(mid_price_tick - bid_depth), hbt.best_bid_tick) * hbt.tick_size
    ask_price = max(np.round(mid_price_tick + ask_depth), hbt.best_ask_tick) * hbt.tick_size
    
    #--------------------------------------------------------
    # Updates quotes.
    
    hbt.clear_inactive_orders()
    
    # Cancel orders if they differ from the updated bid and ask prices.
    for order in hbt.orders.values():
        if order.side == BUY and order.cancellable and order.price != bid_price:
            hbt.cancel(order.order_id)
        if order.side == SELL and order.cancellable and order.price != ask_price:
            hbt.cancel(order.order_id)

    # If the current position is within the maximum position,
    # submit the new order only if no order exists at the same price.
    if hbt.position < max_position and np.isfinite(bid_price):
        bid_price_as_order_id = round(bid_price / hbt.tick_size)
        if bid_price_as_order_id not in hbt.orders:
            hbt.submit_buy_order(bid_price_as_order_id, bid_price, order_qty, GTX)
    if hbt.position > -max_position and np.isfinite(ask_price):
        ask_price_as_order_id = round(ask_price / hbt.tick_size)
        if ask_price_as_order_id not in hbt.orders:
            hbt.submit_sell_order(ask_price_as_order_id, ask_price, order_qty, GTX)
            
    #--------------------------------------------------------
    # Records variables and stats for analysis.
    
    out[t, 0] = half_spread
    out[t, 1] = skew
    out[t, 2] = volatility
    out[t, 3] = A
    out[t, 4] = k
    
    t += 1
    
    if t >= len(arrival_depth) or t >= len(mid_price_chg) or t >= len(out):
        raise Exception
    
    # Records the current state for stat calculation.
    stat.record(hbt)

File: barter-rs/src/strategy/glft.rs
use std::*;

use barter_data::subscription::trade::PublicTrade;
use barter_integration::model::Side;
use ndarray::prelude::*;
use ndarray_stats::{QuantileExt, SummaryStatisticsExt};
use plotpy::{Curve, Plot, StrError};

#[derive(Clone, Debug, Copy)]
pub struct MeasurementParams {
    pub best_bid_tick: f64,
    pub best_ask_tick: f64,
    pub tick_size: f64,
    pub mid_price_tick: f64,
    pub index: usize,
}

impl Default for MeasurementParams {
    fn default() -> Self {
        Self {
            best_bid_tick: f64::NAN,
            best_ask_tick: f64::NAN,
            mid_price_tick: f64::NAN,
            tick_size: 0.01,
            index: 0,
        }
    }
}

pub fn measure_trading_intensity_and_volatility<'a>(
    last_trades: &'a mut Vec<PublicTrade>,
    m: &mut MeasurementParams,
    mut arrival_depth: ArrayViewMut<'a, f64, Ix1>,
    mut mid_price_chg: ArrayViewMut<'a, f64, Ix1>,
) {
    if m.mid_price_tick != f64::NAN {
        let mut depth = -(f64::INFINITY);
        for trade in last_trades {
            let side = trade.side;
            let trade_price_tick = trade.price / m.tick_size;
            if side == Side::Buy {
                depth = array![trade_price_tick - m.mid_price_tick, depth]
                    .max_skipnan()
                    .clone();
            } else {
                depth = array![m.mid_price_tick - trade_price_tick, depth]
                    .max_skipnan()
                    .clone();
            }
        }
        arrival_depth[m.index] = depth;
    }

    let prev_mid_price_tick = m.mid_price_tick;
    m.mid_price_tick = (m.best_bid_tick + m.best_ask_tick) / 2.0;

    mid_price_chg[m.index] = m.mid_price_tick - prev_mid_price_tick;

    m.index += 1;
    if m.index >= arrival_depth.len() || m.index >= mid_price_chg.len() {
        panic!("t >= arrival_depth.len() || t >= mid_price_chg.len()");
    }
}

pub fn measure_trading_intensity(
    arrival_depth: ArrayView<f64, Ix1>,
    mut out: Array<usize, Ix1>,
) -> Array<usize, Ix1> {
    let mut max_tick = 0;
    for depth in arrival_depth {
        if depth.is_infinite() {
            continue;
        }
        let tick = (depth / 0.5).round() as isize - 1;
        if tick < 0 || tick >= out.len() as isize {
            continue;
        }
        out.slice_mut(s![..tick]).mapv_inplace(|x| x + 1);
        max_tick = cmp::max(max_tick, tick);
    }
    out.slice_mut(s![..max_tick]).to_owned()
}

#[allow(non_snake_case)]
pub fn get_params(
    arrival_depth: ArrayView<'_, f64, Ix1>,
    mid_price_chg: ArrayView<'_, f64, Ix1>,
    index: usize,
) -> Result<(f64, f64, f64), StrError> {
    let tmp = Array::from_elem(500, 0);
    let a_depth = arrival_depth.slice(s![index + 1 - 6000..index]);
    let mut lambda_ = measure_trading_intensity(a_depth, tmp).map(|e| *e as f64);
    lambda_ /= 600.0;
    // trim data down to get better fit
    lambda_ = lambda_.slice_move(s![..120]);
    let ticks = Array::range(0.0, lambda_.len() as f64, 1.0) + 0.5;

    let y = lambda_.map(|e| e.ln());
    let (k_, log_a) = linear_regression(&ticks, &y);
    let A = log_a.exp();
    let k = -(k_);

    // configure curve
    // let mut curve = Curve::new();
    // curve.set_line_width(2.0);

    // curve.draw(&ticks.to_vec(), &lambda_.to_vec());

    // let fitted = A * (-k * &ticks).map(|e| e.exp());
    // curve.draw(&ticks.to_vec(), &fitted.to_vec());

    // add curve to plot
    // let mut plot = Plot::new();
    // plot.add(&curve)
    //     .grid_and_labels("delta (ticks from the mid-price)", "Count (per second)");

    // // save figure
    // plot.save("./plot/doc_curve_methods.svg").unwrap();

    // Volatility
    // let firstNan = mid_price_chg.iter().position(|&x| x.is_nan()).unwrap();
    // let mid_price_chg = mid_price_chg.slice_move(s![firstNan + 1..]);
    // let secondNan = mid_price_chg.iter().position(|&x| x.is_nan()).unwrap();
    // let mid_price_chg = mid_price_chg.slice_move(s![..secondNan]);

    let mid_price_chg = mid_price_chg.slice(s![index + 1 - 6000..index]);
    let weights = Array::from_elem(mid_price_chg.len(), 1_f64);

    // mid_price_chg
    //     .iter()
    //     .for_each(|p| println!("mid price: {:?}", p));

    // Since we need volatility in ticks per square root of a second and our measurement is every 100ms,
    // multiply by the square root of 10.
    let volatility = mid_price_chg
        .view()
        .weighted_std(&weights.view(), 0.0)
        .unwrap()
        * 10.0_f64.sqrt();

    Ok((A, k, volatility))
}

fn linear_regression(x: &Array<f64, Ix1>, y: &Array<f64, Ix1>) -> (f64, f64) {
    let sx = x.sum();
    let sy = y.sum();
    let sx2 = x.mapv(|a| a.powi(2)).sum();
    let w = x.len() as f64;
    let sxy = (x * y).sum();
    let slope = ((w * sxy) - (sx * sy)) / ((w * sx2) - sx.powi(2));
    let intercept = (sy - (slope * sx)) / w;
    return (slope, intercept);
}

#[allow(non_snake_case)]
pub fn compute_coeff(xi: f64, gamma: f64, delta: f64, A: f64, k: f64) -> (f64, f64) {
    let inv_k = 1.0 / k;
    let c1 = (1.0 / (xi * delta)) * f64::ln(1.0 + ((xi * delta) * inv_k));
    let c2 = f64::sqrt(
        (gamma / (((2.0 * A) * delta) * k))
            * (1.0 + ((xi * delta) * inv_k)).powf(k / (xi * delta) + 1.0),
    );
    return (c1, c2);
}
